src/base/Group.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <algorithm>
+#include <set>
+#include <boost/iterator/indirect_iterator.hpp>
+
+namespace NuTo
+{
+namespace Groups
+{
+
+template <typename T>
+class Group : private std::set<T*>
+{
+public:
+    typedef std::set<T*> parent;
+    typedef boost::indirect_iterator<typename parent::iterator> GroupIterator;
+
+    void AddMember(T& element)
+    {
+        parent::insert(&element);
+    }
+
+    GroupIterator begin()
+    {
+        return parent::begin();
+    }
+
+    GroupIterator end()
+    {
+        return parent::end();
+    }
+
+    typename parent::iterator pbegin()
+    {
+        return parent::begin();
+    }
+
+    typename parent::iterator pend()
+    {
+        return parent::end();
+    }
+
+    typename parent::const_iterator pcbegin() const
+    {
+        return parent::cbegin();
+    }
+
+    typename parent::const_iterator pcend() const
+    {
+        return parent::cend();
+    }
+
+    bool Contains(const T& element)
+    {
+        auto result = std::find(parent::begin(), parent::end(), &element);
+        if (result != parent::end())
+            return true;
+        else
+            return false;
+    }
+
+    using parent::size;
+    using typename parent::iterator;
+    using typename parent::value_type;
+    using parent::insert;
```

Comment:
Why do we want to allow users to use `insert` and `AddMember`?

Reply:
You are right, we do not want users to use `insert`. However, it is required by the [`insert_iterator`](http://en.cppreference.com/w/cpp/iterator/insert_iterator). 
> The container's insert() member function is called whenever the iterator (whether dereferenced or not) is assigned to.

If we want to keep `Unite(), Difference(),...` as free functions, we have to keep that. As well as all this `pbegin(), pcend()` stuff. And the more I think about it, the more it annoys me as well. Any suggestions? :)